ethers-solc/src/resolver.rs

```diff
@@ -0,0 +1,518 @@
+            })?;
+
+            tracing::trace!("verifying solc checksum for {}", solc.solc.display());
+            if solc.verify_checksum().is_err() {
+                tracing::trace!("corrupted solc version, redownloading  \"{}\"", version);
+                Solc::blocking_install(version.as_ref())?;
+                tracing::trace!("reinstalled solc: \"{}\"", version);
+            }
+            sources_by_version
+                .insert(solc.arg("--allow-paths").arg(allowed_lib_paths.to_string()), sources);
+        }
+        Ok(sources_by_version)
+    }
+}
+
+#[derive(Debug)]
+pub struct Node {
+    path: PathBuf,
+    source: Source,
+    data: SolData,
+}
+
+#[derive(Debug, Clone)]
+#[allow(unused)]
+struct SolData {
+    version: Option<String>,
+    version_req: Option<VersionReq>,
+    imports: Vec<PathBuf>,
+}
+
+fn read_node(file: impl AsRef<Path>) -> Result<Node> {
+    let file = file.as_ref();
+    let source = Source::read(file)?;
+    let data = parse_data(source.as_ref());
+    Ok(Node { path: file.to_path_buf(), source, data })
+}
+
+/// Extracts the useful data from a solidity source
+///
+/// This will attempt to parse the solidity AST and extract the imports and version pragma. If
+/// parsing fails, we'll fall back to extract that info via regex
+fn parse_data(content: &str) -> SolData {
+    let mut version = None;
+    let mut imports = Vec::new();
+    match solang::parser::parse(content, 0) {
+        Ok(units) => {
+            for unit in units.0 {
+                match unit {
+                    SourceUnitPart::PragmaDirective(_, pragma, value) => {
+                        if pragma.name == "solidity" {
+                            // we're only interested in the solidity version pragma
+                            version = Some(value.string);
+                        }
+                    }
+                    SourceUnitPart::ImportDirective(_, import) => {
+                        let import = match import {
+                            Import::Plain(s) => s,
+                            Import::GlobalSymbol(s, _) => s,
+                            Import::Rename(s, _) => s,
+                        };
+                        imports.push(PathBuf::from(import.string));
+                    }
+                    _ => {}
+                }
+            }
+        }
+        Err(err) => {
+            tracing::trace!(
+                "failed to parse solidity ast: \"{:?}\". Falling back to regex to extract data",
+                err
+            );
+            version = utils::find_version_pragma(content).map(str::to_string);
+            imports = utils::find_import_paths(content)
+                .into_iter()
+                .map(|p| Path::new(p).to_path_buf())
+                .collect()
+        }
+    };
+    let version_req = if let Some(ref v) = version { Solc::version_req(v).ok() } else { None };
+    SolData { version_req, version, imports }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
let's add a test to see that we cover the cache issue with libs

Reply:
added test that checks changes in a lib are detected